packages/workspace/src/workspace/path_index.ts

```diff
@@ -189,45 +183,67 @@ export const getTopLevelPathHints = (unmergedElements: Element[]): PathHint[] =>
-  await current.clear()
-  await current.setAll(entries)
+export type PathIndexArgs = {
+  pathIndex: PathIndex
+  unmergedElements: Element[]
+  removedElementsFullNames?: Set<string>
 }
 
-type UpdateIndexParams = {
-  index: PathIndex
-  elements: Element[]
-  accountsToMaintain: string[]
-  isTopLevel: boolean
-}
+/**
+ * Returns all paths from the parent to the element, for example:
+ * 'a.b.c.d' and 'a.b.c.d.e.f.g' will return ['a.b.c.d', 'a.b.c.d.e', 'a.b.c.d.e.f', 'a.b.c.d.e.f.g']
+ */
+const getPathsFromParent = (parentPath: string, elemPath: string): string[] => {
+  // If the parent is not the prefix of the element, it means the element is not a child of the parent
+  if (!elemPath.startsWith(parentPath)) {
+    return []
+  }
 
-export const updatePathIndex = async (
-  { index,
-    elements,
-    accountsToMaintain,
-    isTopLevel }: UpdateIndexParams
+  const parentSegments = parentPath.split(ElemID.NAMESPACE_SEPARATOR)
+  const elemSegments = elemPath.split(ElemID.NAMESPACE_SEPARATOR)
+
+  const subPaths = [parentPath]
+  for (let i = parentSegments.length; i < elemSegments.length - 1; i += 1) {
+    subPaths.push(elemSegments.slice(0, i + 1).join(ElemID.NAMESPACE_SEPARATOR))
+  }
+
+  return subPaths
+}
+/**
+ *  Because currently a change in an element's path doesn't create a change, we are unable to detect it
+ *  We have to override the path index with all the elements, and delete the elements that were removed
+* */
+const updateIndex = async (
+  { pathIndex, unmergedElements, removedElementsFullNames = new Set<string>(), getHintsFunction }:
+    PathIndexArgs &
+    { getHintsFunction: (unmergedElements: Element[]) => RemoteMapEntry<Path[]>[] }
 ): Promise<void> => {
-  if (accountsToMaintain.length === 0) {
-    if (isTopLevel) {
-      await overrideTopLevelPathIndex(index, elements)
-      return
+  const entriesToSet = getHintsFunction(unmergedElements)
+
+  // Entries that are related to an element that was removed should be deleted
+  const entriesToDelete = await awu(pathIndex.keys()).filter(key => {
+    if (removedElementsFullNames.has(key)) {
+      return true
     }
-    await overridePathIndex(index, elements)
-    return
-  }
-  const entries = isTopLevel ? getTopLevelPathHints(elements) : getElementsPathHints(elements)
-  const oldPathHintsToMaintain = await awu(index.entries())
-    .filter(e => accountsToMaintain.includes(ElemID.fromFullName(e.key).adapter))
-    .toArray()
-  const updatedEntries = _.unionBy(entries, oldPathHintsToMaintain, e => e.key)
-  await index.clear()
-  await index.setAll(awu(updatedEntries))
+    // Needed for createTopLevelParentID() function
+    const tempElemID = ElemID.fromFullName(key)
+    const parentFullName = tempElemID.createTopLevelParentID().parent.getFullName()
+    // If any of the levels above the element was removed, delete the element
+    return getPathsFromParent(parentFullName, key).some(path => removedElementsFullNames.has(path))
```

Comment:
You don't have to re-invent the wheel


```suggestion
    const keyElemID = ElemID.fromFullName(key)
    // If any of the levels above the key was removed, delete the key
    return keyElemID.createAllElemIdParents()
      .map(id => id.getFullName())
      .some(fullName => removedElementsFullNames.has(fullName))
```